README.md
Blockwatch R Package
=========

This is Blockwatch's R package. The Blockwatch R package uses the [Blockwatch API](https://blockwatch.cc/docs/api). The official Blockwatch R package manual can be found [here](https://blockwatch.cc/docs/r).

For more information please contact alex@blockwatch.cc

# Installation

To install from Github via the [devtools](https://cran.r-project.org/package=devtools) package:

    install.packages("devtools")
    library(devtools)
    install_github("blockwatch-cc/blockwatch-r")

# Authentication

To make full use of the package we recommend you set your [api key](https://blockwatch.cc/docs/api#getting-started). To do this create or sign into your account and go to your [account api key page](https://blockwatch.cc/account/profile#apikey). Then input your API key (with quotes):

```r
blockwatch.api_key("APIKEY12345678")
```

# Usage

The Blockwatch package functions use the Blockwatch API. Optional Blockwatch API query parameters can be passed into each function. For more information on supported query parameters, please see the [Blockwatch API documentation page](https://blockwatch.cc/docs/api). Once you find the data you would like to load into R on Blockwatch, copy the Blockwatch code from the description box and paste it into the function.

```r
data <- blockwatch.series("BITFINEX:OHLCV/BTC_USD")
```

## Graphing Data Example

To create a graph of Bitcoin's performance month-over-month:

```r
plot(stl(blockwatch.series("BITFINEX:OHLCV/BTC_USD", datatype="ts", collapse="monthly")[,3],s.window="per"))
```

Note: `collapse` is a Blockwatch API query parameter. Click [here](https://blockwatch.cc/docs/api#time-series-api) for a full list of query parameter options.

## Return Types

The supported return types for the `blockwatch.series(code)` function are:
* raw (which returns a data.frame)
* [ts](https://stat.ethz.ch/R-manual/R-devel/library/stats/html/ts.html)
* [zoo](https://cran.r-project.org/package=zoo)
* [xts](https://cran.r-project.org/package=xts)
* [timeSeries](https://cran.r-project.org/package=timeSeries)

To request a specific type, assign the `type` argument the return type:

```r
data <- blockwatch.series('BITFINEX:OHLCV/BTC_USD', datatype = "xts")
```

### Date Formats

zoo, xts, and ts have their own time series date formats. For example:

```r
data <- blockwatch.series('BITFINEX:OHLCV/BTC_USD', collapse = "weekly", datatype = "zoo", limit = 3)
```

`data` will have indexes `2018 Q1`, `2018 Q2`, and `2018 Q3`:

```r
        Open price Close price Highest price Lowest price Average price Trade count Buy count
2018 Q1     6925.3      7508.1        9990.0       6425.1      8236.767     4834496   2358593
2018 Q2     7508.1      7020.6        8506.7       5755.0      6866.010     5112258   2412164
2018 Q3     7020.5      4305.0        7788.0       3657.6      6275.323     4229066   2060333
        Sell count Base volume Quote volume Base volume buy side Quote volume buy side
2018 Q1    2475903     2066609  16842512450            1014616.5            8276759925
2018 Q2    2700094     2405368  16475286016            1133072.0            7771619735
2018 Q3    2168733     1985497  11659640076             931883.9            5459749590
        Base volume sell side Quote volume sell side Standard deviation Mean price
2018 Q1               1051992             8565752525          0.0341565   7508.088
2018 Q2               1272296             8703666281          1.8305689   7023.428
2018 Q3               1053613             6199890485          0.6380526   4306.060
```

If you want the time series index to be displayed as dates, you will need to set `force_irregular = TRUE`:

```r
data <- blockwatch.series('BITFINEX:OHLCV/BTC_USD', collapse = "quarterly", datatype = "zoo", limit = 3, force_irregular = TRUE)
```

`data` will now have indexes `2018-02-28`, `2018-05-30`, and `2018-08-29` (note that interval boundaries for monthly and quartery data aggregates may be imprecise right now):

```r
                    Open price Close price Highest price Lowest price Average price Trade count
2018-02-28 06:00:00     6925.3      7508.1        9990.0       6425.1      8236.767     4834496
2018-05-30 12:00:00     7508.1      7020.6        8506.7       5755.0      6866.010     5112258
2018-08-29 18:00:00     7020.5      4305.0        7788.0       3657.6      6275.323     4229066
                    Buy count Sell count Base volume Quote volume Base volume buy side
2018-02-28 06:00:00   2358593    2475903     2066609  16842512450            1014616.5
2018-05-30 12:00:00   2412164    2700094     2405368  16475286016            1133072.0
2018-08-29 18:00:00   2060333    2168733     1985497  11659640076             931883.9
                    Quote volume buy side Base volume sell side Quote volume sell side
2018-02-28 06:00:00            8276759925               1051992             8565752525
2018-05-30 12:00:00            7771619735               1272296             8703666281
2018-08-29 18:00:00            5459749590               1053613             6199890485
                    Standard deviation Mean price
2018-02-28 06:00:00          0.0341565   7508.088
2018-05-30 12:00:00          1.8305689   7023.428
2018-08-29 18:00:00          0.6380526   4306.060```

## Merged Dataset Data
If you want to get multiple codes at once, delimit the codes with ',', and put them into an array. This will return a multiset.

```r
merged_data <- blockwatch.series(c('BITFINEX:OHLCV/BTC_USD', 'BINANCE:OHLCV/_BTC_USDT'))
```

You can also specify specific columns to retrieve. For example, if you only want column `close` from `BITFINEX:OHLCV/BTC_USD` and column `close` from `PRICE/BINANCE_BTC_USDT`:

```r
merged_data <- blockwatch.series(c('BITFINEX:OHLCV/BTC_USD', 'BINANCE:OHLCV/BTC_USDT'), columns="close")
```

## Downloading Entire Database

An entire database's data cannot be downloaded at the moment. Please check back for updates.

## Datatables

To retrieve Datatable data, provide a Datatable code to the Blockwatch datatables function:

```r
data = blockwatch.table('BITFINEX:TRADE/BTC_USD')
```

The output format is `data.frame`. Given the volume of data stored in datatables, this call will retrieve the first page of the BITFINEX:TRADE/BTC_USD datatable. You may turn on pagination to return more data by using:

```r
data = blockwatch.table('BITFINEX:TRADE/BTC_USD', paginate=TRUE)
```

This will retrieve multiple pages of data and merge them together as if they were one large page. In some cases, however, you will still exceed the request limit. In this case we recommend you filter your data using the available query parameters, as in the following example:

```r
blockwatch.table('BITFINEX:TRADE/BTC_USD', time.gte='2019-01-01', columns='time,price,amount')
```

In this query we are asking for more pages of data, a trade time that is greater than or equal to 2019-01-01 and we are selecting only columns time, price and amount for return rather than all available columns.

## Help

Blockwatch-R can display inline help for databases, datasets and datafields in the R console:

```r
blockwatch.help("BITFINEX:OHLCV")
blockwatch.help("BITFINEX:OHLCV/BTC_USD")
blockwatch.help("BITFINEX:OHLCV/BTC_USD/vwap")
```

In addition you can list subscribed databases, datasets and datafields:

```r
blockwatch.databases()
blockwatch.datasets("BITFINEX:OHLCV")
blockwatch.datafields("BITFINEX:OHLCV/BTC_USD")
```


# Additional Resources

More help can be found at [Blockwatch](https://blockwatch.cc) in our [R](https://blockwatch/docs/r) and [API](https://blockwatch.cc/docs/api) pages.

License provided by MIT. This package is loosely based on https://github.com/quandl/quandl-r created by Raymond McTaggart.


CHANGELOG
### Version 1.0.4

- print embedded result errors
- fix column mapping for empty responses

### Version 1.0.3

- allow setting the base_url
- renamed functions for consistency
- removed unused/unsupported functions

### Version 1.0.2

- removed unused code
- detect column types for table and series responses
- fixed pagination
- add all api metadata responses as dataframe attributes
- append URL query params only when specified

### Version 1.0.0

- forked from Quandl.r package



tests/testthat.R
library(testthat)
test_check("Blockwatch")


tests/testthat/test-api.r
source("test-helpers.r")

context('Test api error handling')

context('blockwatch.api')
with_mock(
  `httr::content` = function(response, as="text") {
    "{}"
  },
  `httr::VERB` = function(http, url, config, body, query) {
    httr:::response(status_code = 500)
  },
  test_that('When status code is not 200 error is thrown', {
    expect_error(blockwatch.api('series'))
  })
)

context('blockwatch.api.download_file')
with_mock(
  `httr::content` = function(response, as="text") {
    "{}"
  },
  `httr::GET` = function(url, config, query, ...) {
    httr:::response(status_code = 403)
  },
  test_that('When status code is not 200 error is thrown', {
    expect_error(blockwatch.api.download_file('series', 'foobar'))
  })
)

context('blockwatch.api.build_request')
test_that('request headers and query params are constructed', {
  Blockwatch.api_key('test_key')
  Blockwatch.api_version('1')
  path <- 'series'
  results <- blockwatch.api.build_request(path, start_date = '2019-01-01', end_date = as.Date('2019-02-01'))
  expected_params <- list(start_date = '2019-01-01', end_date = '2019-02-01')
  expected_headers <- list(
      Accept = 'application/json, application/vnd.blockwatch.cc+json;version=1',
      `Request-Source` = 'R',
      `Request-Source-Version` = '1.0.2',
      `X-Api-Key` = 'test_key'
    )
  expected_url <- "https://data.blockwatch.cc/v1/series"
  expect_equal(results, list(request_url = expected_url, headers = expected_headers, params = expected_params))
})

context('blockwatch.api.build_query_params')
test_that('query params with array values are converted properly', {
  params <- list()
  params$param1 <- 'foo'
  params$param2 <- c('hello', 'world', 'bar')
  params$param3 <- 'cool'

  expected_params <- list(param1='foo', 'param2[]'='hello', 'param2[]'='world', 'param2[]'='bar', param3='cool')
  expect_equal(blockwatch.api.build_query_params(params), expected_params)
})

reset_config()


tests/testthat/test-get.r
library("zoo")
library("xts")
library("timeSeries")
library("httr")
source("test-helpers.r")

context("Getting Dataset data")

context("Blockwatch() bad argument errors")
test_that("Invalid transform throws error", {
  expect_error(Blockwatch("BITFINEX:OHLCV/BTC_USD", transform = "blah"))
})

test_that("Invalid collapse throws error", {
  expect_error(Blockwatch("BITFINEX:OHLCV/BTC_USD", collapse = "blah"))
})

test_that("Invalid type throws error", {
  expect_error(Blockwatch("BITFINEX:OHLCV/BTC_USD", datatype = "blah"))
})

context("Blockwatch() call")
with_mock(
  `httr::VERB` = function(http, url, config, body, query) {
    print(query)
    test_that("correct arguments are passed in", {
      expect_equal(http, "GET")
      expect_equal(url, "https://data.blockwatch.cc/v1/series/BITFINEX:OHLCV/BTC_USD.json")
      expect_null(body)
      expect_equal(query, list(start_date = "2019-01-01", transform = "rdiff",
                               collapse = "annual", order = "desc"))
    })
    mock_response()
  },
  `httr::content` = function(response, as="text") {
    response$content
  },
  Blockwatch("BITFINEX:OHLCV/BTC_USD", start_date = "2019-01-01", transform = "rdiff", collapse = "annual", order = "desc")
)

context("Blockwatch() daily data response")
with_mock(
  `httr::VERB` = function(http, url, config, body, query) {
    mock_response()
  },
  `httr::content` = function(response, as="text") {
    response$content
  },
  test_that("list names are set to column names", {
    dataset <- Blockwatch("BITFINEX:OHLCV/BTC_USD")
    expect_named(dataset, c("Timestamp", "Open price", "Close price", "Highest price",
      "Lowest price", "Average price", "Trade count", "Buy count", "Sell count",
      "Base volume", "Quote volume", "Base volume buy side", "Quote volume buy side",
      "Base volume sell side", "Quote volume sell side", "Standard deviation", "Mean price"))
  }),
  test_that("returned data is a dataframe", {
    dataset <- Blockwatch("BITFINEX:OHLCV/BTC_USD")
    expect_is(dataset, 'data.frame')
  }),
  # test_that("does contain meta attribute when requested", {
  #   dataset <- Blockwatch("BITFINEX:OHLCV/BTC_USD", meta = TRUE)
  #   expect_true(!is.null(attr(dataset, "meta")))
  #   expect_equal(attr(dataset, "meta")$id, 6668)
  #   expect_equal(attr(dataset, "meta")$database_code, "NSE")
  #   expect_equal(attr(dataset, "meta")$dataset_code, "OIL")
  # }),
  test_that("zoo is returned when requested", {
    dataset <- Blockwatch("BITFINEX:OHLCV/BTC_USD", datatype = "zoo")
    expect_is(dataset, "zoo")
  }),
  test_that("xts is returned when requested", {
    dataset <- Blockwatch("BITFINEX:OHLCV/BTC_USD", datatype = "xts")
    expect_is(dataset, "xts")
  }),
  test_that("timeSeries is returned when requested", {
    dataset <- Blockwatch("BITFINEX:OHLCV/BTC_USD", datatype = "timeSeries")
    expect_is(dataset, "timeSeries")
  }),
  test_that("zoo is returned instead of ts if ts is not supported for frequency", {
    dataset <- Blockwatch("BITFINEX:OHLCV/BTC_USD", datatype = "ts")
    expect_is(dataset, "zoo")
  }),
  test_that("display warning message if type ts is not supported by frequency", {
    expect_warning(Blockwatch("BITFINEX:OHLCV/BTC_USD", datatype = "ts"),
      "Type 'ts' does not support frequency 365. Returning zoo.", fixed = TRUE)
  })
)

context("Blockwatch() annual collapse response data")
with_mock(
  `httr::VERB` = function(http, url, config, body, query) {
    mock_response(content = mock_annual_data())
  },
  `httr::content` = function(response, as = "text") {
    response$content
  },
  test_that("return ts when requested", {
    dataset <- Blockwatch("BITFINEX:OHLCV/BTC_USD", datatype = "ts", collapse = "annual")
    expect_is(dataset, "ts")
  }),
  test_that("Data is the same across formats", {
    annaulraw <- Blockwatch("BITFINEX:OHLCV/BTC_USD", datatype="raw", collapse = "annual")
    annaults <- Blockwatch("BITFINEX:OHLCV/BTC_USD", datatype="ts", collapse = "annual")
    annaulzoo <- Blockwatch("BITFINEX:OHLCV/BTC_USD", datatype="zoo", collapse = "annual")
    annaulxts <- Blockwatch("BITFINEX:OHLCV/BTC_USD", datatype="xts", collapse = "annual")
    annaultimeSeries <- Blockwatch("BITFINEX:OHLCV/BTC_USD", datatype="timeSeries")
    expect_equal(max(abs(annaults - coredata(annaulzoo))), 0)
    expect_equal(max(abs(coredata(annaulzoo) - coredata(annaulxts))) , 0)
    # timeSeries keeps data in same order as passed in, not chronological
    # Have to compare against raw as zoo and xts are sorted chronologically
    expect_equal(max(abs(annaulraw[,-1] - getDataPart(annaultimeSeries))), 0)
  }),
  test_that("When requesting xts annual warn user xts has non-standard meaning for frequency", {
    expect_warning(Blockwatch("BITFINEX:OHLCV/BTC_USD", datatype="xts", collapse = "annual"),
      "xts has a non-standard meaning for 'frequency'.", fixed = TRUE)
  })
)

# FIXME
# context("Blockwatch() monthly collapse response data")
# with_mock(
#   `httr::VERB` = function(http, url, config, body, query) {
#     mock_response(content = mock_monthly_data())
#   },
#   `httr::content` = function(response, as = "text") {
#     response$content
#   },
#   test_that("Frequencies are correct across output formats", {
#     monthlyts <- Blockwatch("BITFINEX:OHLCV/BTC_USD", datatype="ts", collapse = "monthly")
#     monthlyzoo <- Blockwatch("BITFINEX:OHLCV/BTC_USD", datatype="zoo", collapse = "monthly")
#     monthlyxts <- Blockwatch("BITFINEX:OHLCV/BTC_USD", datatype="xts", collapse = "monthly")
#     monthlytimeSeries <- Blockwatch("BITFINEX:OHLCV/BTC_USD", datatype="timeSeries", collapse = "monthly")

#     expect_equal(frequency(monthlyts), 12)
#     expect_equal(frequency(monthlyzoo), 12)
#     expect_equal(frequency(monthlyxts), 12)
#     # timeSeries allows time index in reverse order but regularity checks won't work then
#     # So we check reversed series also
#     expect_true((frequency(monthlytimeSeries)==12)||(frequency(rev(monthlytimeSeries))==12))
#   })
# )

# FIXME
# context("Blockwatch() annual frequency response data")
# with_mock(
#   `httr::VERB` = function(http, url, config, body, query) {
#     mock_response(content = mock_annual_frequency_data())
#   },
#   `httr::content` = function(response, as = "text") {
#     response$content
#   },
#   test_that("Start and end dates are correct (zoo)", {
#     annual <- Blockwatch("BITFINEX:OHLCV/BTC_USD", datatype="zoo")
#     expect_equal(start(annual), 2015)
#     expect_equal(end(annual), 2018)
#   }),
#     test_that("Start and end dates are correct (zoo) with force_irregular param set", {
#     annual <- Blockwatch("BITFINEX:OHLCV/BTC_USD", datatype="zoo", force_irregular = TRUE)
#     expect_equal(start(annual), as.Date("2015-05-01"))
#     expect_equal(end(annual), as.Date("2018-12-31"))
#   }),
#   test_that("Start and end dates are correct (xts)", {
#     annual <- Blockwatch("TESTS/4", datatype="xts")
#     expect_that(start(annual), is_equivalent_to(as.Date("2015-05-01")))
#     expect_that(end(annual), is_equivalent_to(as.Date("2018-12-31")))
#   }),
#   test_that("Start and end dates are correct (timeSeries)", {
#    annual <- Blockwatch("TESTS/4", datatype="timeSeries")
#    expect_that(start(annual), is_equivalent_to(as.timeDate("2015-05-01")))
#    expect_that(end(annual), is_equivalent_to(as.timeDate("2018-12-31")))
#   })
# )

context("Blockwatch() quarterly frequency response data")
with_mock(
  `httr::VERB` = function(http, url, config, body, query) {
    mock_response(content = mock_quarterly_collapse_data())
  },
  `httr::content` = function(response, as = "text") {
    response$content
  },
  test_that("Collapsed data frequency", {
    dailytoquart <- Blockwatch("TESTS/1", datatype="ts", collapse="quarterly")
    expect_equal(frequency(dailytoquart), 4)
  })
)

# context('Blockwatch() filtering by column index')
# with_mock(
#   `httr::VERB` = function(http, url, config, body, query) {
#     test_that("correct arguments are passed in", {
#       expect_equal(query$column, "id")
#     })
#     mock_response()
#   },
#   `httr::content` = function(response, as="text") {
#     response$content
#   },
#   Blockwatch("BITFINEX:OHLCV/BTC_USD.id")
# )

# context('Blockwatch() multiple datasets')
# test_that("Multiple dataset codes with dataset code column indexes are requested", {
#   database_codes <- c("NSE", "AAPL")
#   dataset_codes <- c("OIL", "WIKI")
#   requested_column_indexes <- c("1", "2")
#   i <- 0
#   with_mock(
#     `httr::VERB` = function(http, url, config, body, query) {
#       i <<- i + 1
#       test_that("request is made with correct params", {
#         expected_code <- paste(database_codes[i], dataset_codes[i], sep = "/")
#         expect_true(grepl(expected_code, url))
#         expect_equal(query$column_index, requested_column_indexes[i])
#         expect_equal(query$transform, "rdiff")
#       })
#       mock_response(content = mock_data(database_code = database_codes[i], dataset_code = dataset_codes[i]))
#     },
#     `httr::content` = function(response, as = "text") {
#       response$content
#     },
#     Blockwatch(c("BITFINEX:OHLCV/BTC_USD.1", "AAPL/WIKI.2"), transform = "rdiff", column_index = 3)
#   )
#   expect_equal(i, 2)
# })

# test_that("Multiple dataset codes calls merge", {
#   database_codes <- c("NSE", "AAPL", "TESTS")
#   dataset_codes <- c("OIL", "WIKI", "4")
#   i <- 0
#   with_mock(
#     `httr::VERB` = function(http, url, config, body, query) {
#       mock_response(content = mock_data(database_code = database_codes[i], dataset_code = dataset_codes[i]))
#     },
#     `httr::content` = function(response, as = "text") {
#       response$content
#     },
#     merge = function(data, merge_data, ...) {
#       i <<- i + 1
#     },
#     Blockwatch(c("BITFINEX:OHLCV/BTC_USD.1", "WIKI/AAPL.2", "TESTS/4.1"))
#   )
#   expect_equal(i, 2, info = "3 datasets are merged together, merged is called twice")
# })

# test_that("Multiple dataset codes returns desired requested datatype", {
#   with_mock(
#     `httr::VERB` = function(http, url, config, body, query) {
#       mock_response(content = mock_annual_data())
#     },
#     `httr::content` = function(response, as = "text") {
#       response$content
#     },
#     test_that("return datatype is correct", {
#       datatypes <- c('raw', 'ts', 'zoo', 'xts', 'timeSeries')
#       expected <- c('data.frame', 'ts', 'zoo', 'xts', 'timeSeries')
#       i <- 0
#       for(datatype in datatypes) {
#         i <- i + 1
#         dataset = Blockwatch(c("BITFINEX:OHLCV/BTC_USD.1", "WIKI/AAPL.2", "TESTS/4.1"), datatype = datatype, collapse = 'annual')
#         expect_is(dataset, expected[i])
#       }
#     })
#   )
# })

reset_config()


R/Blockwatchtable.R
#' Retrieves Data from a Blockwatch Datatable endpoint
#'
#' @details Set your \code{api_key} with \code{blockwatch.api_key} function. For instructions on finding your api key go to \url{https://blockwatch.cc/accountprofile#/apikey}
#'
#' @param table_code Datatable code on Blockwatch specified as a string.
#' @param paginate When set to TRUE, fetches up to 1,000,000 rows of data
#' @param ... Additional named values that are interpreted as Blockwatch API parameters.
#' @return Returns a data.frame.
#' @seealso \code{\link{blockwatch.api_key}}
#' @examples \dontrun{
#' blockwatch.table('BITFINEX:TRADE/BTC_USD', paginate=TRUE)
#' }
#' @export
blockwatch.table <- function(table_code, paginate = FALSE, ...) {
  path <- paste0("tables/", table_code)
  params <- list(...)

  # when pagination is on, fetch at max limit
  if (isTRUE(paginate) && is.null(params$limit)) {
    params["limit"] <- blockwatch.api.max_limit()
  }

  # make request for first page of data
  json <- do.call(blockwatch.api, c(path = path, params))

  # check embedded error
  if (!is.null(json$error)) {
    stop(paste(jsonlite::toJSON(json$error, auto_unbox=TRUE, pretty=TRUE)), call. = FALSE)
  }

  df <- as.data.frame(json$data, stringsAsFactors = FALSE)
  cursor <- json$cursor

  # continue to make requests for data if paginate=TRUE and there is data
  while (isTRUE(paginate) && length(json$data) > 0) {
    params["cursor"] <- cursor
    json <- do.call(blockwatch.api, c(path = path, params))
    if (!is.null(json$error)) {
      stop(paste(jsonlite::toJSON(json$error, auto_unbox=TRUE, pretty=TRUE)), call. = FALSE)
    }
    df_page <- as.data.frame(json$data, stringsAsFactors = FALSE)

    df <- rbind(df, df_page)
    cursor <- json$cursor

    # only fetch a maximum of 1,000,000 rows
    if (blockwatch.table.max_rows() > 0 && nrow(df) >= blockwatch.table.max_rows() && !is.null(cursor)) {
      if (!getOption("blockwatch.quiet", FALSE)) {
        warning(paste("This call returns more data than R can process efficiently.",
                      "See https://blockwatch.cc/docs/r#usage-limits for details."), call. = FALSE)
      }
      break
    }
  }

  if (!isTRUE(paginate) && !is.null(cursor) && json$limit == json$count) {
    if (!getOption("blockwatch.quiet", FALSE)) {
      warning(paste("This call can return more data. Increase your limit or set paginate=TRUE.",
                    "See https://blockwatch.cc/docs/r#accessing-tables-with-r for details."), call. = FALSE)
    }
  }

  # if df is empty create an empty df with ncolumns set
  # or else we won't be able to set the column names
  ncols <- length(json$columns$code)
  if (nrow(df) <= 0 && ncols > 0) {
    df <- data.frame(matrix(ncol = ncols, nrow = 0))
  }

  # set column names
  names(df) <- json$columns$code

  # set column types
  column_types <- tolower(json$columns$type)
  for (i in 1:length(column_types)) {
    if (grepl("^float|^bigdecimal|^int|^uint", column_types[i])) {
      df[, i] <- as.numeric(df[, i])
    } else if (grepl("^datetime", column_types[i])) {
      df[, i] <- as.POSIXct(as.numeric(df[, i])/1000, origin = "1970-01-01", tz="UTC")
    } else if (grepl("^date", column_types[i])) {
      df[, i] <- as.Date(df[, i])
    } else if (grepl("^bool", column_types[i])) {
      df[, i] <- as.logical(as.numeric(df[, i]))
    } else {
      df[, i] <- as.character(df[, i])
    }
  }

  return(df)
}


R/Blockwatchseries.R
#' Retrieves Data from a Blockwatch Time Series endpoint
#'
#' @details Set your \code{api_key} with \code{blockwatch.api_key} function. For instructions on finding your api key go to \url{https:/blockwatch.cc/account/profile#apikey}.
#'
#' @param code Dataset code on Blockwatch specified as a string or an array of strings.
#' @param datatype Type of data returned specified as string. Can be 'raw', 'ts', 'zoo', 'xts' or 'timeSeries'.
#' @param transform Apply Blockwatch API data transformations.
#' @param collapse Collapse frequency of Data.
#' @param order Select if data is given to R in ascending or descending formats. Helpful for the rows parameter.
#' @param force_irregular When set to TRUE, forces the index of the Data to be of date format yyyy-mm-dd
#' @param ... Additional named values that are interpreted as Blockwatch API parameters. Please see \url{https:/blockwatch.cc/docs/api#database-metadata-api} for a full list of parameters.
#' @return Depending on the type the class is either data.frame, time series, xts, zoo or timeSeries.
#' @references This R package uses the Blockwatch API. For more information go to \url{https:/blockwatch.cc/docs/api}. For more help on the package itself go to \url{https:/blockwatch.cc/docs/r}.
#' @seealso \code{\link{blockwatch.api_key}}
#' @examples \dontrun{
#' data = blockwatch.series("BITFINEX:OHLCV/BTC_USD", start_date="2018-01-01", datatype="ts")
#' plot(data[,1])
#' }
#' @importFrom zoo zoo
#' @importFrom zoo as.zooreg
#' @importFrom zoo as.yearmon
#' @importFrom zoo as.yearqtr
#' @importFrom xts xts
#' @importFrom xts as.xts
#' @export
blockwatch.series <- function(code, datatype = c("raw", "ts", "zoo", "xts", "timeSeries"),
                   transform = c("", "diff", "rdiff", "normalize", "cumul", "rdiff_from"),
                   collapse = c("", "1m", "5m", "15m", "30m", "1h", "3h", "6h", "12h", "1d", "1w", "1M", "3M", "1y", "daily", "weekly", "monthly", "quarterly", "annual"),
                   order = c("asc", "desc"), force_irregular = FALSE, ...) {
  params <- list(...)
  ## Check params
  datatype <- match.arg(datatype)
  if (!is.null(transform) && transform != "") {
    params$transform <- match.arg(transform)
  }
  if (!is.null(collapse) && collapse != "") {
    params$collapse <- match.arg(collapse)
  }
  if (!is.null(order) && order != "") {
    params$order <- match.arg(order)
  }

  ## validate date format if supplied
  if (!is.null(params$start_date)) {
    as.Date(params$start_date)
  }

  if (!is.null(params$end_date)) {
    as.Date(params$end_date)
  }

  if (datatype == "timeSeries" && system.file(package = datatype) == "") {
    stop("Package ", datatype, " needed to use this datatype", call. = FALSE)
  }

  ## Helper functions
  frequency2integer <- function(freq) {
    if (is.null(freq) || is.na(freq)) {
      return(365)
    } else {
      switch(freq,
           # "1m", "5m", "15m", "30m", "1h", "3h", "6h", "12h",
           # cannot express as integers with seasonality = 1 year!
           # "1d", "1w", "1M", "3M", "1y"
           "daily"     = 365,
           "1d"        = 365,
           "weekly"    = 52,
           "1w"        = 52,
           "monthly"   = 12,
           "1M"        = 12,
           "quarterly" = 4,
           "3M"        = 4,
           "yearly"    = 1,
           "1y"        = 1,
           1)
    }
  }

  as.year <- function(x) {
    floor(as.numeric(as.yearmon(x)))
  }

  check.code <- function(code) {
    if (!all(gsub("[^A-Z0-9_.:/\\-]", "", code) == code)) {
      stop("Dataset codes are comprised of capital letters, numbers, dashes and underscores only.")
    }

    codearray <- strsplit(code, "/")
    if (length(codearray[[1]]) != 2) {
      stop("Invalid dataset code! Must be of form <database_code>/<dataset_code>")
    }
  }

  freq <- frequency2integer(params$collapse)

  ## Download and parse data
  errors <- list()
  if (length(code) == 1) {
    # fetch a single series
    check.code(code)
    data <- blockwatch.series.get(code, params)
  } else {
    # fetch multiple series and merge
    data <- NULL
    for (c in code) {
      tmp.params <- params
      check.code(c)

      merge_data <- tryCatch(blockwatch.series.get(c, tmp.params), error = function(e) {
        d <- data.frame(Date = character(0), ERROR = numeric(0))
        attr(d, "errors") <- e
        return(d)
      })

      suppressWarnings(errors[c] <- attr(merge_data, "errors"))

      for (i in 2:length(names(merge_data))) {
        names(merge_data)[i] <- paste(sub("/", ".", c), names(merge_data)[i], sep = " - ")
      }

      if (is.null(data)) {
        data <- merge_data
      } else {
        data <- merge(data, merge_data, by = 1, all = TRUE)
      }
    }
  }

  ## Returning raw data
  if (datatype == "raw") {
    data_out <- data
  } else {
    # Deal with regularly spaced time series first
    if (freq %in% c(1, 4, 12) && !force_irregular) {
      # Build regular zoo with correct frequency
      if (freq == 1) {
        data_out <- zoo::zoo(data[, -1], frequency = freq, as.year(data[, 1]))
      } else if (freq == 4) {
        data_out <- zoo::zoo(
          data[, -1], frequency = freq, as.yearqtr(data[, 1])
        )
      } else if (freq == 12) {
        data_out <- zoo::zoo(
          data[, -1], frequency = freq, as.yearmon(data[, 1])
        )
      }

      # Convert to type
      if (datatype == "ts") {
        data_out <- stats::as.ts(data_out)
      } else if (datatype == "zoo") {
        data_out <- zoo::as.zooreg(data_out)
      } else if (datatype == "xts") {
        data_out <- if (freq == 1) {
          xts::xts(data[, -1], frequency = 1, order.by = data[, 1])
        } else  {
          xts::as.xts(data_out)
        }
        if (freq != stats::frequency(data_out)) {
          warning("xts has a non-standard meaning for 'frequency'.")
        }
      } else if (datatype == "timeSeries") {
        data_out <- timeSeries::timeSeries(
          data = data[, -1],
          charvec = data[, 1]
        )
      }

    } else if (datatype == "zoo" || datatype == "ts") {
      # Time series is not regularly spaced
      if (datatype == "ts") {
        warning("Type 'ts' does not support frequency ", freq, ". Returning zoo.")
      }
      data_out <- zoo::zoo(data[, -1], order.by = data[, 1])
    } else if (datatype == "xts") {
      data_out <- xts::xts(data[, -1], order.by = data[, 1])
    } else if (datatype == "timeSeries") {
      data_out <- timeSeries::timeSeries(data = data[, -1], charvec = data[, 1])
    }
  }

  if (length(errors) > 0) {
    attr(data_out, "errors") <- errors
  }

  return(data_out)
}

#' Internal method to retrieve data from a Blockwatch time series endpoint
#'
#' @param code Dataset code on Blockwatch specified as a string.
#' @param params A list of parameters to be passed to the Blockwatch API. Please see \url{https:/blockwatch.cc/docs/api#retrieve-data-and-metadata} for a full list of parameters.
#' @return Returns a data.frame of the requested data
#' @seealso \code{\link{blockwatch.api_key}}, \code{\link{blockwatch.table}}
#' @examples \dontrun{
#' data = blockwatch.series.get("TRADE/BITFINEX_BTC_USD", list(limit=5))
#' plot(data[,1])
#' }
blockwatch.series.get <- function(code, params) {
  path <- paste0("series/", code)
  json <- do.call(blockwatch.api, c(path = path, params))

  if (!is.null(json$error)) {
    stop(paste(jsonlite::toJSON(json$error, auto_unbox=TRUE, pretty=TRUE)), call. = FALSE)
  }

  if (length(json$data) == 0) {
    # if df is empty create an empty df with ncolumns set
    # or else we won't be able to set the column names
    ncols <- length(json$columns)
    if (ncols > 0) {
      df <- data.frame(matrix(ncol = ncols, nrow = 0))
      warning("empty server response")
    } else {
      stop("Error: empty server response", )
    }
  } else {
    df <- as.data.frame(json$data, stringsAsFactors = FALSE)
  }
  names(df) <- json$columns$code

  ## set columns types (transform values to numeric, date or string)
  column_types <- tolower(json$columns$type)
  for (i in 1:length(column_types)) {
    if (grepl("^float|^bigdecimal|^int|^uint", column_types[i])) {
      df[, i] <- as.numeric(df[, i])
    } else if (grepl("^datetime", column_types[i])) {
      df[, i] <- as.POSIXct(as.numeric(df[, i])/1000, origin = "1970-01-01", tz="UTC")
    } else if (grepl("^date", column_types[i])) {
      df[, i] <- as.Date(df[, i])
    } else if (grepl("^bool", column_types[i])) {
      df[, i] <- as.logical(as.numeric(df[, i]))
    } else {
      df[, i] <- as.character(df[, i])
    }
  }

  # keep extra result parameters as attributes
  attr(df, "transform") <- json$transform
  attr(df, "collapse") <- json$collapse
  attr(df, "order") <- json$order
  attr(df, "start_date") <- as.POSIXct(json$start_date/1000, origin = "1970-01-01", tz="UTC")
  attr(df, "end_date") <- as.POSIXct(json$end_date/1000, origin = "1970-01-01", tz="UTC")
  attr(df, "limit") <- json$limit
  attr(df, "count") <- json$count

  return(df)
}
